extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/CdcSource.java

```diff
@@ -0,0 +1,248 @@
+            return 0L;
+        } else {
+            return ((Struct) record.value()).getInt64("__ts_ms");
+        }
+    }
+
+    private class JetSourceTaskContext implements SourceTaskContext {
+        @Override
+        public Map<String, String> configs() {
+            return taskConfig;
+        }
+
+        @Override
+        public OffsetStorageReader offsetStorageReader() {
+            return new SourceOffsetStorageReader();
+        }
+    }
+
+    private class SourceOffsetStorageReader implements OffsetStorageReader {
+        @Override
+        public <V> Map<String, Object> offset(Map<String, V> partition) {
+            return offsets(Collections.singletonList(partition)).get(partition);
+        }
+
+        @Override
+        public <V> Map<Map<String, V>, Map<String, Object>> offsets(Collection<Map<String, V>> partitions) {
+            Map<Map<String, V>, Map<String, Object>> map = new HashMap<>();
+            for (Map<String, V> partition : partitions) {
+                Map<String, Object> offset = (Map<String, Object>) state.getOffset(partition);
+                map.put(partition, offset);
+            }
+            return map;
+        }
+    }
+
+    private static class JetConnectorContext implements ConnectorContext {
+        @Override
+        public void requestTaskReconfiguration() {
+            // no-op since it is not supported
+        }
+
+        @Override
+        public void raiseError(Exception e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static final class State implements IdentifiedDataSerializable {
+
+        /**
+         * Key represents the partition which the record originated from. Value
+         * represents the offset within that partition. Kafka Connect represents
+         * the partition and offset as arbitrary values so that is why it is
+         * stored as map.
+         * See {@link SourceRecord} for more information regarding the format.
+         */
+        private final Map<Map<String, ?>, Map<String, ?>> partitionsToOffset = new HashMap<>();
+        private final List<byte[]> historyRecords = new CopyOnWriteArrayList<>();
+
+        public Map<String, ?> getOffset(Map<String, ?> partition) {
+            return partitionsToOffset.get(partition);
+        }
+
+        public void setOffset(Map<String, ?> partition, Map<String, ?> offset) {
+            partitionsToOffset.put(partition, offset);
+        }
+
+        public List<byte[]> getHistoryRecords() {
+            return historyRecords;
+        }
+
+        @Override
+        public int getFactoryId() {
+            return CdcJsonDataSerializerHook.FACTORY_ID;
+        }
+
+        @Override
+        public int getClassId() {
+            return CdcJsonDataSerializerHook.SOURCE_STATE;
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+            out.writeObject(partitionsToOffset);
+
+            List<byte[]> records = historyRecords;
```

Comment:
you can use writeObject to write a List.

Reply:
For some reason I though this would be more space-efficient, but I should have checked. Will revisit.